File: examples/crypto/sha256/test-vectors.zig
const std = @import("std");
const crypto = @import("crypto");
const SHA256 = crypto.SHA256;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("=== SHA256 NIST Test Vectors ===\n\n", .{});
    std.debug.print("Running NIST FIPS 180-4 test vectors...\n\n", .{});

    var passed: u32 = 0;
    var failed: u32 = 0;

    // Test 1: Empty string
    {
        std.debug.print("Test: Empty string\n", .{});
        std.debug.print("{s}\n", .{"-" ** 70});

        const input = "";
        const expected = [_]u8{
            0xe3, 0xb0, 0xc4, 0x42, 0x98, 0xfc, 0x1c, 0x14,
            0x9a, 0xfb, 0xf4, 0xc8, 0x99, 0x6f, 0xb9, 0x24,
            0x27, 0xae, 0x41, 0xe4, 0x64, 0x9b, 0x93, 0x4c,
            0xa4, 0x95, 0x99, 0x1b, 0x78, 0x52, 0xb8, 0x55,
        };

        const hash = SHA256.hash(input);
        const is_match = std.mem.eql(u8, &hash, &expected);

        std.debug.print("Input:    \"{s}\"\n", .{input});
        std.debug.print("Expected: 0x{s}\n", .{std.fmt.fmtSliceHexLower(&expected)});
        std.debug.print("Got:      0x{s}\n", .{std.fmt.fmtSliceHexLower(&hash)});
        std.debug.print("Status:   {s}\n\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});

        if (is_match) passed += 1 else failed += 1;
    }

    // Test 2: "abc"
    {
        std.debug.print("Test: abc\n", .{});
        std.debug.print("{s}\n", .{"-" ** 70});

        const input = "abc";
        const expected = [_]u8{
            0xba, 0x78, 0x16, 0xbf, 0x8f, 0x01, 0xcf, 0xea,
            0x41, 0x41, 0x40, 0xde, 0x5d, 0xae, 0x22, 0x23,
            0xb0, 0x03, 0x61, 0xa3, 0x96, 0x17, 0x7a, 0x9c,
            0xb4, 0x10, 0xff, 0x61, 0xf2, 0x00, 0x15, 0xad,
        };

        const hash = SHA256.hash(input);
        const is_match = std.mem.eql(u8, &hash, &expected);

        std.debug.print("Input:    \"{s}\"\n", .{input});
        std.debug.print("Expected: 0x{s}\n", .{std.fmt.fmtSliceHexLower(&expected)});
        std.debug.print("Got:      0x{s}\n", .{std.fmt.fmtSliceHexLower(&hash)});
        std.debug.print("Status:   {s}\n\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});

        if (is_match) passed += 1 else failed += 1;
    }

    // Test 3: "hello world"
    {
        std.debug.print("Test: hello world\n", .{});
        std.debug.print("{s}\n", .{"-" ** 70});

        const input = "hello world";
        const expected = [_]u8{
            0xb9, 0x4d, 0x27, 0xb9, 0x93, 0x4d, 0x3e, 0x08,
            0xa5, 0x2e, 0x52, 0xd7, 0xda, 0x7d, 0xab, 0xfa,
            0xc4, 0x84, 0xef, 0xe3, 0x7a, 0x53, 0x80, 0xee,
            0x90, 0x88, 0xf7, 0xac, 0xe2, 0xef, 0xcd, 0xe9,
        };

        const hash = SHA256.hash(input);
        const is_match = std.mem.eql(u8, &hash, &expected);

        std.debug.print("Input:    \"{s}\"\n", .{input});
        std.debug.print("Expected: 0x{s}\n", .{std.fmt.fmtSliceHexLower(&expected)});
        std.debug.print("Got:      0x{s}\n", .{std.fmt.fmtSliceHexLower(&hash)});
        std.debug.print("Status:   {s}\n\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});

        if (is_match) passed += 1 else failed += 1;
    }

    // Test 4: 448-bit message
    {
        std.debug.print("Test: 448-bit message\n", .{});
        std.debug.print("{s}\n", .{"-" ** 70});

        const input = "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq";
        const expected = [_]u8{
            0x24, 0x8d, 0x6a, 0x61, 0xd2, 0x06, 0x38, 0xb8,
            0xe5, 0xc0, 0x26, 0x93, 0x0c, 0x3e, 0x60, 0x39,
            0xa3, 0x3c, 0xe4, 0x59, 0x64, 0xff, 0x21, 0x67,
            0xf6, 0xec, 0xed, 0xd4, 0x19, 0xdb, 0x06, 0xc1,
        };

        const hash = SHA256.hash(input);
        const is_match = std.mem.eql(u8, &hash, &expected);

        std.debug.print("Input:    \"{s}\"\n", .{input});
        std.debug.print("Expected: 0x{s}\n", .{std.fmt.fmtSliceHexLower(&expected)});
        std.debug.print("Got:      0x{s}\n", .{std.fmt.fmtSliceHexLower(&hash)});
        std.debug.print("Status:   {s}\n\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});

        if (is_match) passed += 1 else failed += 1;
    }

    // Test 5: 896-bit message
    {
        std.debug.print("Test: 896-bit message\n", .{});
        std.debug.print("{s}\n", .{"-" ** 70});

        const input = "abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmnoijklmnopjklmnopqklmnopqrlmnopqrsmnopqrstnopqrstu";
        const expected = [_]u8{
            0xcf, 0x5b, 0x16, 0xa7, 0x78, 0xaf, 0x83, 0x80,
            0x03, 0x6c, 0xe5, 0x9e, 0x7b, 0x04, 0x92, 0x37,
            0x0b, 0x24, 0x9b, 0x11, 0xe8, 0xf0, 0x7a, 0x51,
            0xaf, 0xac, 0x45, 0x03, 0x7a, 0xfe, 0xe9, 0xd1,
        };

        const hash = SHA256.hash(input);
        const is_match = std.mem.eql(u8, &hash, &expected);

        std.debug.print("Input:    \"{s}...\" ({} bytes)\n", .{ input[0..47], input.len });
        std.debug.print("Expected: 0x{s}\n", .{std.fmt.fmtSliceHexLower(&expected)});
        std.debug.print("Got:      0x{s}\n", .{std.fmt.fmtSliceHexLower(&hash)});
        std.debug.print("Status:   {s}\n\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});

        if (is_match) passed += 1 else failed += 1;
    }

    // Test 6: One million 'a' characters
    {
        std.debug.print("Test: One million \"a\" characters\n", .{});
        std.debug.print("{s}\n", .{"-" ** 70});

        const expected = [_]u8{
            0xcd, 0xc7, 0x6e, 0x5c, 0x99, 0x14, 0xfb, 0x92,
            0x81, 0xa1, 0xc7, 0xe2, 0x84, 0xd7, 0x3e, 0x67,
            0xf1, 0x80, 0x9a, 0x48, 0xa4, 0x97, 0x20, 0x0e,
            0x04, 0x6d, 0x39, 0xcc, 0xc7, 0x11, 0x2c, 0xd0,
        };

        std.debug.print("Input:    1,000,000 × \"a\"\n", .{});
        std.debug.print("Expected: 0x{s}\n", .{std.fmt.fmtSliceHexLower(&expected)});

        const timer = try std.time.Timer.start();

        // Use streaming API for memory efficiency
        var hasher = SHA256.init();
        const chunk_size = 10000;
        var chunk = try allocator.alloc(u8, chunk_size);
        defer allocator.free(chunk);
        @memset(chunk, 'a');

        var i: usize = 0;
        while (i < 100) : (i += 1) {
            hasher.update(chunk);
        }

        const hash = hasher.final();
        const elapsed = timer.read();

        const is_match = std.mem.eql(u8, &hash, &expected);

        std.debug.print("Got:      0x{s}\n", .{std.fmt.fmtSliceHexLower(&hash)});
        std.debug.print("Time:     {d:.2} ms\n", .{@as(f64, @floatFromInt(elapsed)) / 1_000_000.0});
        std.debug.print("Status:   {s}\n\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});

        if (is_match) passed += 1 else failed += 1;
    }

    // Edge case tests
    std.debug.print("Edge Case Tests\n", .{});
    std.debug.print("{s}\n", .{"-" ** 70});

    // Single byte (0x00)
    {
        const input = [_]u8{0x00};
        const expected = [_]u8{
            0x6e, 0x34, 0x0b, 0x9c, 0xff, 0xb3, 0x7a, 0x98,
            0x9c, 0xa5, 0x44, 0xe6, 0xbb, 0x78, 0x0a, 0x2c,
            0x78, 0x90, 0x1d, 0x3f, 0xb3, 0x37, 0x38, 0x76,
            0x85, 0x11, 0xa3, 0x06, 0x17, 0xaf, 0xa0, 0x1d,
        };
        const hash = SHA256.hash(&input);
        const is_match = std.mem.eql(u8, &hash, &expected);
        std.debug.print("Single byte (0x00): {s}\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});
        if (is_match) passed += 1 else failed += 1;
    }

    // All zeros (32 bytes)
    {
        const input = [_]u8{0x00} ** 32;
        const expected = [_]u8{
            0x66, 0x68, 0x7a, 0xad, 0xf8, 0x62, 0xbd, 0x77,
            0x6c, 0x8f, 0xc1, 0x8b, 0x8e, 0x9f, 0x8e, 0x20,
            0x08, 0x97, 0x14, 0x85, 0x6e, 0xe2, 0x33, 0xb3,
            0x90, 0x2a, 0x59, 0x1d, 0x0d, 0x5f, 0x29, 0x25,
        };
        const hash = SHA256.hash(&input);
        const is_match = std.mem.eql(u8, &hash, &expected);
        std.debug.print("All zeros (32 bytes): {s}\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});
        if (is_match) passed += 1 else failed += 1;
    }

    // All ones (32 bytes)
    {
        const input = [_]u8{0xFF} ** 32;
        const expected = [_]u8{
            0xaf, 0x96, 0x13, 0x76, 0x0f, 0x72, 0x63, 0x5f,
            0xbd, 0xb4, 0x4a, 0x5a, 0x0a, 0x63, 0xc3, 0x9f,
            0x12, 0xaf, 0x30, 0xf9, 0x50, 0xa6, 0xee, 0x5c,
            0x97, 0x1b, 0xe1, 0x88, 0xe8, 0x9c, 0x40, 0x51,
        };
        const hash = SHA256.hash(&input);
        const is_match = std.mem.eql(u8, &hash, &expected);
        std.debug.print("All ones (32 bytes): {s}\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});
        if (is_match) passed += 1 else failed += 1;
    }

    std.debug.print("\n", .{});

    // Unicode tests
    std.debug.print("Unicode Tests\n", .{});
    std.debug.print("{s}\n", .{"-" ** 70});

    // Emoji
    {
        const input = "🚀";
        const bytes = [_]u8{ 0xF0, 0x9F, 0x9A, 0x80 };
        const string_hash = SHA256.hash(input);
        const bytes_hash = SHA256.hash(&bytes);
        const is_match = std.mem.eql(u8, &string_hash, &bytes_hash);
        std.debug.print("Emoji 🚀: {s}\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});
        std.debug.print("  UTF-8 bytes: 0x{s}\n", .{std.fmt.fmtSliceHexLower(&bytes)});
        std.debug.print("  Hash:        0x{s}...\n", .{std.fmt.fmtSliceHexLower(string_hash[0..20])});
        if (is_match) passed += 1 else failed += 1;
    }

    // Chinese
    {
        const input = "你好";
        const bytes = [_]u8{ 0xE4, 0xBD, 0xA0, 0xE5, 0xA5, 0xBD };
        const string_hash = SHA256.hash(input);
        const bytes_hash = SHA256.hash(&bytes);
        const is_match = std.mem.eql(u8, &string_hash, &bytes_hash);
        std.debug.print("Chinese 你好: {s}\n", .{if (is_match) "✓ PASS" else "✗ FAIL"});
        std.debug.print("  UTF-8 bytes: 0x{s}\n", .{std.fmt.fmtSliceHexLower(&bytes)});
        std.debug.print("  Hash:        0x{s}...\n", .{std.fmt.fmtSliceHexLower(string_hash[0..20])});
        if (is_match) passed += 1 else failed += 1;
    }

    std.debug.print("\n", .{});

    // Summary
    std.debug.print("{s}\n", .{"=" ** 70});
    std.debug.print("Test Summary\n", .{});
    std.debug.print("{s}\n", .{"=" ** 70});
    std.debug.print("Total tests:  {}\n", .{passed + failed});
    std.debug.print("Passed:       {} ✓\n", .{passed});
    std.debug.print("Failed:       {}\n", .{failed});

    if (failed == 0) {
        std.debug.print("\n🎉 All tests passed! Implementation is correct.\n", .{});
    } else {
        std.debug.print("\n⚠️  Some tests failed. Check implementation.\n", .{});
    }

    std.debug.print("\n=== Test Vectors Complete ===\n", .{});
}

File: examples/crypto/sha256/test-vectors.ts
/**
 * SHA256 NIST Test Vectors
 *
 * Validates implementation against official NIST FIPS 180-4 test vectors:
 * - Empty string
 * - Short messages
 * - Messages at block boundaries
 * - Large messages (1 million 'a' characters)
 */

import { SHA256 } from '../../../src/crypto/sha256/SHA256.js';

console.log('=== SHA256 NIST Test Vectors ===\n');

interface TestVector {
  name: string;
  input: string;
  expected: string;
}

const testVectors: TestVector[] = [
  {
    name: 'Empty string',
    input: '',
    expected: '0xe3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
  },
  {
    name: 'abc',
    input: 'abc',
    expected: '0xba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad',
  },
  {
    name: 'hello world',
    input: 'hello world',
    expected: '0xb94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9',
  },
  {
    name: '448-bit message',
    input: 'abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq',
    expected: '0x248d6a61d20638b8e5c026930c3e6039a33ce45964ff2167f6ecedd419db06c1',
  },
  {
    name: '896-bit message',
    input: 'abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmnoijklmnopjklmnopqklmnopqrlmnopqrsmnopqrstnopqrstu',
    expected: '0xcf5b16a778af8380036ce59e7b0492370b249b11e8f07a51afac45037afee9d1',
  },
];

console.log('Running NIST FIPS 180-4 test vectors...\n');

let passed = 0;
let failed = 0;

for (const vector of testVectors) {
  console.log(`Test: ${vector.name}`);
  console.log('-'.repeat(70));

  const hash = SHA256.hashString(vector.input);
  const result = SHA256.toHex(hash);

  const isMatch = result === vector.expected;

  if (vector.input.length <= 50) {
    console.log('Input:   ', `"${vector.input}"`);
  } else {
    console.log('Input:   ', `"${vector.input.slice(0, 47)}..." (${vector.input.length} bytes)`);
  }
  console.log('Expected:', vector.expected);
  console.log('Got:     ', result);
  console.log('Status:  ', isMatch ? '✓ PASS' : '✗ FAIL');

  if (isMatch) {
    passed++;
  } else {
    failed++;
  }

  console.log();
}

// Special test: One million 'a' characters
console.log('Test: One million "a" characters');
console.log('-'.repeat(70));

const largeInput = 'a'.repeat(1000000);
const largeExpected = '0xcdc76e5c9914fb9281a1c7e284d73e67f1809a48a497200e046d39ccc7112cd0';

console.log('Input:    1,000,000 × "a"');
console.log('Expected:', largeExpected);

const start = performance.now();
const largeHash = SHA256.hashString(largeInput);
const elapsed = performance.now() - start;

const largeResult = SHA256.toHex(largeHash);
const isLargeMatch = largeResult === largeExpected;

console.log('Got:     ', largeResult);
console.log('Time:    ', elapsed.toFixed(2), 'ms');
console.log('Status:  ', isLargeMatch ? '✓ PASS' : '✗ FAIL');

if (isLargeMatch) {
  passed++;
} else {
  failed++;
}

console.log();

// Test with streaming API (should produce same result)
console.log('Test: Streaming API validation');
console.log('-'.repeat(70));

const hasher = SHA256.create();
const chunkSize = 10000;
for (let i = 0; i < 100; i++) {
  hasher.update(new TextEncoder().encode('a'.repeat(chunkSize)));
}
const streamingHash = hasher.digest();
const streamingResult = SHA256.toHex(streamingHash);

console.log('Input:    1,000,000 × "a" (streamed in 10KB chunks)');
console.log('Expected:', largeExpected);
console.log('Got:     ', streamingResult);
console.log('Status:  ', streamingResult === largeExpected ? '✓ PASS' : '✗ FAIL');

if (streamingResult === largeExpected) {
  passed++;
} else {
  failed++;
}

console.log();

// Edge case tests
console.log('Edge Case Tests');
console.log('-'.repeat(70));

const edgeCases = [
  {
    name: 'Single byte (0x00)',
    input: new Uint8Array([0x00]),
    expected: '0x6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d',
  },
  {
    name: 'All zeros (32 bytes)',
    input: new Uint8Array(32).fill(0x00),
    expected: '0x66687aadf862bd776c8fc18b8e9f8e20089714856ee233b3902a591d0d5f2925',
  },
  {
    name: 'All ones (32 bytes)',
    input: new Uint8Array(32).fill(0xFF),
    expected: '0xaf9613760f72635fbdb44a5a0a63c39f12af30f950a6ee5c971be188e89c4051',
  },
];

for (const test of edgeCases) {
  const hash = SHA256.hash(test.input);
  const result = SHA256.toHex(hash);
  const isMatch = result === test.expected;

  console.log(`${test.name}:`, isMatch ? '✓ PASS' : '✗ FAIL');
  if (!isMatch) {
    console.log('  Expected:', test.expected);
    console.log('  Got:     ', result);
  }

  if (isMatch) {
    passed++;
  } else {
    failed++;
  }
}

console.log();

// Unicode tests
console.log('Unicode Tests');
console.log('-'.repeat(70));

const unicodeTests = [
  {
    name: 'Emoji 🚀',
    input: '🚀',
    // UTF-8: F0 9F 9A 80
    bytes: new Uint8Array([0xF0, 0x9F, 0x9A, 0x80]),
  },
  {
    name: 'Chinese 你好',
    input: '你好',
    // UTF-8: E4 BD A0 E5 A5 BD
    bytes: new Uint8Array([0xE4, 0xBD, 0xA0, 0xE5, 0xA5, 0xBD]),
  },
];

for (const test of unicodeTests) {
  const stringHash = SHA256.hashString(test.input);
  const bytesHash = SHA256.hash(test.bytes);
  const isMatch = SHA256.toHex(stringHash) === SHA256.toHex(bytesHash);

  console.log(`${test.name}:`, isMatch ? '✓ PASS' : '✗ FAIL');
  console.log('  UTF-8 bytes:', Array.from(test.bytes).map(b => '0x' + b.toString(16).padStart(2, '0')).join(' '));
  console.log('  Hash:       ', SHA256.toHex(stringHash).slice(0, 40) + '...');

  if (isMatch) {
    passed++;
  } else {
    failed++;
  }
}

console.log();

// Summary
console.log('='.repeat(70));
console.log('Test Summary');
console.log('='.repeat(70));
console.log('Total tests: ', passed + failed);
console.log('Passed:      ', passed, '✓');
console.log('Failed:      ', failed, failed > 0 ? '✗' : '');

if (failed === 0) {
  console.log('\n🎉 All tests passed! Implementation is correct.');
} else {
  console.log('\n⚠️  Some tests failed. Check implementation.');
}

console.log();
console.log('=== Test Vectors Complete ===');
